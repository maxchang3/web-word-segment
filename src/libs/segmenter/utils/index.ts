export * from './regex'
export * from './text'
export interface Segmenter {
    isReady: boolean
    init(...args: any[]): Promise<void>
    segment(text: string): Promise<string[]>
}
