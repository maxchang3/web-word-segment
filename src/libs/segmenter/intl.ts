import { canBeSkip, type Segmenter, splitTextByURL } from './utils'

class IntlSegmenter implements Segmenter {
    isReady: boolean
    tool: Intl.Segmenter
    async init() {
        this.isReady = true
    }

    constructor() {
        this.tool = new Intl.Segmenter('zh', { granularity: 'word' })
        this.isReady = false
    }

    *segmentIter(text: string): Iterable<string> {
        text = text.trim()
        if (text === '') {
            return
        }
        const textSplitByURL = splitTextByURL(text)
        for (const part of textSplitByURL) {
            if (part.isURL) {
                yield part.text
                continue
            }
            for (const segment of this.tool.segment(part.text)) {
                const word = segment.segment
                if (!canBeSkip(word)) yield word
            }
        }
    }

    async segment(text: string): Promise<string[]> {
        return Array.from(this.segmentIter(text))
    }
}

export const intlSegmenter = new IntlSegmenter()
