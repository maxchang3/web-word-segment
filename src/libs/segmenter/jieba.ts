import { canBeSkip, type Segmenter, splitTextByURL } from './utils'

class JiebaSegmenter implements Segmenter {
    isReady: boolean = false
    cut: ((text: string) => Promise<string[]>) | undefined

    async init() {
        let { default: entry, cut } = await import('jieba-wasm')
        if (typeof entry !== 'function') { // for Node.js & Testing
            // @ts-expect-error for Node.js
            cut = entry.cut
        } else {
            await entry()
        }
        this.cut = async (text: string) => cut(text) as string[]
        this.isReady = true
    }

    async segment(text: string): Promise<string[]> {
        if (!this.isReady) {
            throw new Error('JiebaSegmenter not ready')
        }
        text = text.trim()
        if (text === '') {
            return []
        }
        const textSplitByURL = splitTextByURL(text)
        const result = (await Promise.all( // Cut text and URL
            textSplitByURL.map(async ({ text, isURL }) => isURL ? text : this.cut!(text)),
        ))
            .flat()
            .filter(
                (word: string) => !(canBeSkip(word)),
            )
        return result
    }
}

export const jiebaSegmenter = new JiebaSegmenter()
