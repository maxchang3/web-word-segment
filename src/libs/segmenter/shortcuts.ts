import { intlSegmenter } from './intl'

export * from './utils'

export const supportBackends = ['intl'] as const

export type SupportBackend = typeof supportBackends[number]

export const isSupportIntlSegmenter = typeof Intl?.Segmenter === 'function'

export const backendData = {
    intl: ['Intl.Segmenter', 'https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Intl/Segmenter'],
} satisfies Record<SupportBackend, [string, string]>

export const useSegmenter = (_: SupportBackend = 'intl') => intlSegmenter
