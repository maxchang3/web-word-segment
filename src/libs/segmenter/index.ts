import { intlSegmenter } from './intl'
import { jiebaSegmenter } from './jieba'

export * from './utils'

export const isSupportIntlSegmenter = typeof Intl?.Segmenter === 'function'

export const useSegmenter = (backend: 'jieba' | 'intl' = 'jieba') => {
    if (backend === 'intl') {
        if (isSupportIntlSegmenter) return intlSegmenter
        throw new Error('Intl.Segmenter not supported')
    }
    return jiebaSegmenter
}
