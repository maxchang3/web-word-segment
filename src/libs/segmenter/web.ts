import { intlSegmenter } from './intl'
import { jiebaSegmenter } from './jieba'

export * from './utils'

export const supportBackends = ['jieba', 'intl'] as const

export type SupportBackend = typeof supportBackends[number]

export const isSupportIntlSegmenter = typeof Intl?.Segmenter === 'function'

export const backendData = {
    jieba: ['jieba-wasm', 'https://github.com/fengkx/jieba-wasm'],
    intl: ['Intl.Segmenter', 'https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Intl/Segmenter'],
} satisfies Record<SupportBackend, [string, string]>

export const useSegmenter = (backend: SupportBackend = 'jieba') => {
    if (backend === 'intl') {
        if (isSupportIntlSegmenter) return intlSegmenter
        throw new Error('Intl.Segmenter not supported')
    }
    return jiebaSegmenter
}
