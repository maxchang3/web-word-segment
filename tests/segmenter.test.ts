import { splitTextByURL, supportBackends, useSegmenter } from '@/libs/segmenter/web'
import { expect, it } from 'vitest'

for (const backend of supportBackends) {
    const segmenter = useSegmenter(backend)

    await segmenter.init()

    it('should correctly segment Chinese text', async () => {
        const result = await segmenter.segment('中华人民共和国武汉市长江大桥')
        expect(result).toMatchSnapshot()
    })

    it('should correctly segment English text', async () => {
        const result = await segmenter.segment('Hello, World!')
        expect(result).toMatchSnapshot()
    })

    it('should correctly segment mixed text', async () => {
        const result = await segmenter.segment(`1972年，在贝尔实验室成员布莱恩·柯林汉撰写的内部技术文件《A Tutorial Introduction to the Language B》中首次提到了 Hello World 这一字符串。`)
        expect(result).toMatchSnapshot()
    })

    it('should correctly split text by URL', async () => {
        const result = splitTextByURL('访问我们的官网example.com时，如果您有任何问题，请发送邮件至service@example.com或通过论坛test.googl联系管理员')
        expect(result).toMatchSnapshot()
    })

    it('should correctly segment text with URL and email', async () => {
        const result = await segmenter.segment('访问我们的官网example.com时，如果您有任何问题，请发送邮件至service@example.com或通过论坛test.googl联系管理员')
        expect(result).toMatchSnapshot()
    })
}
