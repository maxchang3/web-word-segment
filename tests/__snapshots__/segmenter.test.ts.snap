// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should correctly segment Chinese text 1`] = `
[
  "中华人民共和国",
  "武汉市",
  "长江大桥",
]
`;

exports[`should correctly segment Chinese text 2`] = `
[
  "中华",
  "人民",
  "共和国",
  "武汉",
  "市",
  "长江",
  "大",
  "桥",
]
`;

exports[`should correctly segment English text 1`] = `
[
  "Hello",
  "World",
]
`;

exports[`should correctly segment English text 2`] = `
[
  "Hello",
  "World",
]
`;

exports[`should correctly segment mixed text 1`] = `
[
  "1972",
  "年",
  "在",
  "贝尔实验室",
  "成员",
  "布莱恩",
  "柯林",
  "汉",
  "撰写",
  "的",
  "内部",
  "技术",
  "文件",
  "A",
  "Tutorial",
  "Introduction",
  "to",
  "the",
  "Language",
  "B",
  "中",
  "首次",
  "提到",
  "了",
  "Hello",
  "World",
  "这一",
  "字符串",
]
`;

exports[`should correctly segment mixed text 2`] = `
[
  "1972",
  "年",
  "在",
  "贝尔",
  "实验",
  "室",
  "成员",
  "布",
  "莱",
  "恩",
  "柯",
  "林",
  "汉",
  "撰写",
  "的",
  "内部",
  "技术",
  "文件",
  "A",
  "Tutorial",
  "Introduction",
  "to",
  "the",
  "Language",
  "B",
  "中",
  "首次",
  "提到",
  "了",
  "Hello",
  "World",
  "这",
  "一",
  "字符",
  "串",
]
`;

exports[`should correctly segment text with URL and email 1`] = `
[
  "访问",
  "我们",
  "的",
  "官网",
  "example.com",
  "时",
  "如果",
  "您",
  "有",
  "任何",
  "问题",
  "请",
  "发送",
  "邮件",
  "至",
  "service@example.com",
  "或",
  "通过",
  "论坛",
  "test.googl",
  "联系",
  "管理员",
]
`;

exports[`should correctly segment text with URL and email 2`] = `
[
  "访问",
  "我们",
  "的",
  "官",
  "网",
  "example.com",
  "时",
  "如果",
  "您",
  "有",
  "任何",
  "问题",
  "请",
  "发送",
  "邮件",
  "至",
  "service@example.com",
  "或",
  "通过",
  "论坛",
  "test.googl",
  "联系",
  "管理",
  "员",
]
`;

exports[`should correctly split text by URL 1`] = `
[
  {
    "isURL": false,
    "text": "访问我们的官网",
  },
  {
    "isURL": true,
    "text": "example.com",
  },
  {
    "isURL": false,
    "text": "时，如果您有任何问题，请发送邮件至",
  },
  {
    "isURL": true,
    "text": "service@example.com",
  },
  {
    "isURL": false,
    "text": "或通过论坛",
  },
  {
    "isURL": true,
    "text": "test.googl",
  },
  {
    "isURL": false,
    "text": "联系管理员",
  },
]
`;

exports[`should correctly split text by URL 2`] = `
[
  {
    "isURL": false,
    "text": "访问我们的官网",
  },
  {
    "isURL": true,
    "text": "example.com",
  },
  {
    "isURL": false,
    "text": "时，如果您有任何问题，请发送邮件至",
  },
  {
    "isURL": true,
    "text": "service@example.com",
  },
  {
    "isURL": false,
    "text": "或通过论坛",
  },
  {
    "isURL": true,
    "text": "test.googl",
  },
  {
    "isURL": false,
    "text": "联系管理员",
  },
]
`;
